File: Data Structures & Algorithms/valid-tree/submission-3.java
class Solution {
    List<List<Integer>> adj = new ArrayList<>();
    Set<Integer> visit = new HashSet<>();
    public boolean dfs(int node,int par){
        if(visit.contains(node)) return false ;
        visit.add(node);
        for(int i : adj.get(node)){
            if(i==par) continue ;
            if(!dfs(i,node)) return false ;
        }
        return true ;
    }
    public boolean validTree(int n, int[][] edges) {
        if(edges.length>n-1) return false ;
        for (int i = 0; i < n; i++) {
            adj.add(new ArrayList<>());
        }

        for(int[] p : edges){
            adj.get(p[0]).add(p[1]);
            adj.get(p[1]).add(p[0]);
        }
        return dfs(0,-1) && visit.size()==n;
    }
}

File: Data Structures & Algorithms/count-connected-components/submission-3.py
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:   
        D={}
        for p in edges : 
            if p[0] not in D :
                D[p[0]]={p[1]}
            else :
                D[p[0]].add(p[1])
            if p[1] not in D :
                D[p[1]]={p[0]}
            else :
                D[p[1]].add(p[0])
        visited=set()
        N=0
        def dfs(n) :
            visited.add(n)
            for x in D[n] :
                if x in D and x not in visited :
                    dfs(x)
                else :
                    visited.add(x)
            return
        for key in D :
            if key not in visited :
                dfs(key)
                N+=1 
        return N+n-len(visited)